src/lib/pubkey/classic_mceliece/cmce_matrix.h

```diff
@@ -0,0 +1,114 @@
+       */
+      static std::optional<std::pair<Classic_McEliece_Matrix, Column_Selection>> create_matrix(
+         const Classic_McEliece_Parameters& params,
+         const Classic_McEliece_Field_Ordering& field_ordering,
+         const Classic_McEliece_Minimal_Polynomial& g);
+
+      /**
+       * @brief Create the matrix H for a Classic McEliece instance given its
+       * parameters, field ordering and minimal polynomial.
+       *
+       * Output is a pair of the matrix and the pivot vector c that was used to
+       * create it in the semi-systematic form as described in Classic McEliece ISO
+       * Section 9.2.11.
+       *
+       * This method directly updates the field ordering values as described in Classic McEliece
+       * ISO Section 7.2.3 Step 5 (for f parameter sets).
+       *
+       * @param params Classic McEliece parameters
+       * @param field_ordering Field ordering (will be updated)
+       * @param g Minimal polynomial
+       * @return Pair(the matrix H, pivot vector c)
+       */
+      static std::optional<std::pair<Classic_McEliece_Matrix, Column_Selection>> create_matrix_and_apply_pivots(
+         const Classic_McEliece_Parameters& params,
+         Classic_McEliece_Field_Ordering& field_ordering,
+         const Classic_McEliece_Minimal_Polynomial& g);
+
+      /**
+       * @brief The bytes of the submatrix T, with H=(I_mt, T) as defined in Classic
+       * McEliece ISO Section 9.2.7.
+       *
+       * @return The matrix bytes
+       */
+      std::vector<uint8_t> bytes() const { return m_mat_bytes; }
+
+      /**
+       * @brief Create a Classic_McEliece_Matrix from bytes.
+       *
+       * @param mat_bytes The bytes of the submatrix T as defined in Classic McEliece ISO Section 9.2.7.
+       */
+      Classic_McEliece_Matrix(const Classic_McEliece_Parameters& params, std::vector<uint8_t> mat_bytes) :
+            m_mat_bytes(std::move(mat_bytes)) {
+         BOTAN_ARG_CHECK(m_mat_bytes.size() == params.pk_size_bytes(), "Invalid byte size for matrix");
+         if(params.pk_no_cols() % 8 == 0) {
+            return;
+         }
+         // Check padding of mat_bytes rows
+         BOTAN_ASSERT_NOMSG(m_mat_bytes.size() == params.pk_no_rows() * params.pk_row_size_bytes());
+         for(size_t row = 0; row < params.pk_no_rows(); ++row) {
+            uint8_t padded_byte = m_mat_bytes[(row + 1) * params.pk_row_size_bytes() - 1];
+            BOTAN_ARG_CHECK(padded_byte >> (params.pk_no_cols() % 8) == 0, "Valid padding of unused bytes");
+         }
+      }
+
+      /**
+       * @brief Multiply the Classic McEliece matrix H with a bitvector e.
+       *
+       * @param params Classic McEliece parameters
+       * @param e The bitvector e
+       * @return H*e
+       */
+      Code_Word mul(const Classic_McEliece_Parameters& params, const Error_Vector& e) const;
+
+   private:
+      /// The bytes of the submatrix T
+      const std::vector<uint8_t> m_mat_bytes;  // can we use bitvector?
```

Comment:
The matrix is the public key of CMCE, i.e. the >1 MB of data. In multiplication, for example, we use it row by row, so we would need the _bitspan_ to avoid copying it around for each encapsulation if we use one huge bitvector. As an alternative, we could store the rows of the matrix like `std::vector<bitvector> m_mat_rows`. This could also speed up consecutive encapsulations, since we do not need to recompute the bitvectors every time. A disadvange could be that the data is not necessarily consecutive in memory this way, which could slow down the memory accesses...